SCRY-517 Scrypto Compiler crate

mstrug-rdx opened the pull request:
## Summary
New `scrypto-compiler` crate.

## Details
Implementation according to [SCRY-517](https://radixdlt.atlassian.net/browse/SCRY-517).
New `ScryptoCompiler` is used in `simulator` and `scrypto-test` projects.

## Testing
Added new tests to validate compilation with `ScryptoCompiler`.



[SCRY-517]: https://radixdlt.atlassian.net/browse/SCRY-517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

0xOmarA reviewed (changes requested):
I think that there are some items that we should address prior to merging the PR.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+        self
+    }
+
+    pub fn force_local_target(&mut self, local_target: bool) -> &mut Self {
```
It was used in `simulator/.../cargo.rs` so I've implemented it on compiler level but you are right that `target_directory` is enough.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+        Ok(())
+    }
+
+    fn prepare_features(&self) -> String {
```
I've used:
```
let features: Vec<&str> = self.input_params.features.iter().map(|f| ["--features", f]).flatten().collect();
```
as we need to specify arguments each as separate string and also no `String` allocation due to use of `&str`.

iamyulong commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+        self.build()?.compile()
+    }
+
+    pub fn compile_workspace(&mut self) -> Result<Vec<PathBuf>, ScryptoCompilerError> {
```
Also can we return `Map<String, PathBuf>`?

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+    // Returns output wasm file path
+    pub fn compile(&mut self) -> Result<PathBuf, ScryptoCompilerError> {
+        // Create compilation command
+        let mut command = Command::new("cargo");
```
I've implemented this by additional compile function: `compile_with_stdio` because `Stdio` object is not implementing `Clone`, so I cannot store it in `InputParams` struct.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+        if input_params.coverage && input_params.target_directory.is_some() {
+            return Err(ScryptoCompilerInvalidInputParam::CoverageDiscardsTargetDirectory);
+        }
+        if input_params.force_local_target && input_params.target_directory.is_some() {
```
I've already removed that. It was taken from simulator compiler implementation, but we don't need that currenlty.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+            .is_some_and(|v| PathBuf::from(v).ends_with(MANIFEST_FILE))
+        {
+            return Err(ScryptoCompilerInvalidInputParam::CargoTomlInManifestDirectory);
+        }
```
Done.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+    // Internal constructor
+    fn from_input_params(
+        input_params: &ScryptoCompilerInputParams,
+    ) -> Result<Self, ScryptoCompilerError> {
```
It is made owned later in that function when it creates `Self`, so in the place where it needs to be coped. Before that there is couple function calls which also gets `input_params` as argument (also reference) so if some of these functions will return error we will exit from `from_input_params` function early without need to copy `input_params`. So I think it is some optimisation (although compiler will do it better on its own probably).

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+    EnvironmentVariableSetAndUnset(String),
+}
+
+#[derive(Clone, Default)]
```
`Default` trait on `ScryptoCompilerInputParams` is used only internally when creating `ScryptoCompilerBuilder` through `ScryptoCompiler::builder()` function. It just initialises Vectors and Options to empty values, only `Profile` is initialised to `Release` also by `Default` trait.
`ScryptoCompilerInputParams` cannot be initialised by the user to use with `ScryptoCompiler`/`ScryptoCompilerBuilder` so I think it should be ok to use `Default` here.

mstrug-rdx commented on scrypto-compiler/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+
+[dependencies]
+radix-engine-interface = { path = "../radix-engine-interface", default-features = false }
+radix-engine = { path = "../radix-engine", default-features = false }
```
I think it is not so straight forward because `extract_definition` internally uses some modules from `radix-engine` to execute wasm file. For instant it uses: `WasmModule`, `WasmRuntime`, `SystemLoanFeeReserve`.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+            .arg(&self.manifest_path)
+            .args(package)
+            .args(features)
+            .env("CARGO_ENCODED_RUSTFLAGS", rustflags)
```
`CARGO_ENCODED_RUSTFLAGS` was used in `scrypto build` command implementation so I just used it here, but I think it is not needed to add special handling of `CARGO_ENCODED_RUSTFLAGS` or `RUSTFLAGS` as they can be provided in `ScryptoCompiler::env()` function now.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+        self.build()?.compile()
+    }
+
+    pub fn compile_workspace(&mut self) -> Result<Vec<PathBuf>, ScryptoCompilerError> {
```
I've reimplemented `workspace` compilation, now it is based on `cargo`.
For item 2, I've implemented it that way it will be compiled only if member of a workspace has defined `[package.metadata.scrypto]` section in its `Cargo.toml` file. For packages specified with `package` input parameter this is not checked.

mstrug-rdx commented on scrypto-compiler/src/lib.rs:
```diff
@@ -0,0 +1,910 @@
+    }
+
+    pub fn coverage(&mut self, coverage: bool) -> &mut Self {
+        self.input_params.coverage = coverage;
```
Moved to coverage command.

mstrug-rdx commented:
I’ve implemented two-phase compilation internally in `ScryptoCompiler` as advised on PR comments.
Also I've updated `scrypto build` command to support all input parameters of `ScryptoCompiler`. Command `resim publish` also builds the package so the question if it should also expose all `ScryptoCompiler` options like `scrypto build`? Currently I left it without change.
1st phase compilation for package definition extraction always uses `Release` compilation profile because on `Debug` there was an error `ExtractSchema(InvalidWasm(TooManyFunctionLocals { max: 256, actual: 552 }))`, so we cannot execute it on `Debug`.

iamyulong reviewed (approved):
Looks really nice!

Finally we now have a unified, consistent Scrypto compilation experience. 

It's more than a refactor, but loads of functionalities being added to the compiler too.

iamyulong commented:
@0xOmarA - Based on a conversation with @mstrug-rdx,  all of the comments have been addressed. 

I'm going to merge the PR, so I can publish them all to crates.io.

Please take a look at the PR again after you're back.
